src/purge.rs

```diff
@@ -66,19 +78,46 @@ where
             &mut will_force_compact,
         );
 
-        let min_file_num = self.memtables.fold(u64::MAX, |min, t| {
-            cmp::min(min, t.min_file_num(LogQueue::Append).unwrap_or(u64::MAX))
-        });
-        let purged = self.pipe_log.purge_to(LogQueue::Append, min_file_num)?;
-        info!("purged {} expired log files", purged);
+        let rewrite_operations = self.global_stats.rewrite_operations();
+        let compacted_rewrite_operations = self.global_stats.compacted_rewrite_operations();
+        if compacted_rewrite_operations as f64 / rewrite_operations as f64 > 0.75
+            && self.needs_purge_log_files(LogQueue::Rewrite)
+        {
+            self.squeeze_rewrite_queue();
+        }
+
+        let (min_file_1, min_file_2) =
+            self.memtables
+                .fold((u64::MAX, u64::MAX), |(mut min1, mut min2), t| {
+                    min1 = cmp::min(min1, t.min_file_num(LogQueue::Append).unwrap_or(u64::MAX));
+                    min2 = cmp::min(min2, t.min_file_num(LogQueue::Rewrite).unwrap_or(u64::MAX));
+                    (min1, min2)
+                });
+        let purged_1 = self.pipe_log.purge_to(LogQueue::Append, min_file_1)?;
+        info!("purged {} expired log files", purged_1);
 
+        if min_file_2 < self.pipe_log.active_file_num(LogQueue::Rewrite) {
+            let purged_2 = self.pipe_log.purge_to(LogQueue::Rewrite, min_file_2)?;
+            info!("purged {} expired rewrite files", purged_2);
+        }
+
+        drop(purge_mutex);
         Ok(will_force_compact)
     }
 
-    pub fn needs_purge_log_files(&self) -> bool {
-        let total_size = self.pipe_log.total_size(LogQueue::Append);
+    pub fn needs_purge_log_files(&self, queue: LogQueue) -> bool {
+        let active_file_num = self.pipe_log.active_file_num(queue);
+        let first_file_num = self.pipe_log.first_file_num(queue);
+        if active_file_num == first_file_num {
+            return false;
+        }
+
+        let total_size = self.pipe_log.total_size(queue);
         let purge_threshold = self.cfg.purge_threshold.0 as usize;
-        total_size > purge_threshold
+        match queue {
+            LogQueue::Append => total_size > purge_threshold,
+            LogQueue::Rewrite => total_size * 10 > purge_threshold,
```

Comment:
It is two small.... This means that we only allow 10% of data is rewritten? Maybe we could do some test and then adjust this ratio.

Reply:
It's used to reduce the frequence of squeeze the rewrite queue. The rewrite queue can only be squeezed when the garbage ratio is greater than 75%, and the total log file size of the rewrite queue reaches 1GB (by default).